File: ios/Plugin/HmsCapacitorPlugin.swift
import Foundation
import Capacitor
import HMSSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(HmsCapacitorPlugin)
public class HmsCapacitorPlugin: CAPPlugin {
    private let implementation = HmsCapacitor()
    
    var hmsSDK: HMSSDK!;
    
    var hmsConfig: HMSConfig!;
    
    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        let audioState: String
        
        switch AVAudioSession.sharedInstance().recordPermission{
        case AVAudioSession.RecordPermission.granted:
            audioState = "granted"
        case AVAudioSession.RecordPermission.denied:
            audioState = "denied"
        case AVAudioSession.RecordPermission.undetermined:
            audioState = "undetermined"
        
        default:
            audioState = "prompt"
        }
        
        call.resolve(["record_audio": audioState])
    }
    
    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        AVAudioSession.sharedInstance().requestRecordPermission { granted in
            if granted {
                // The user granted access. Present recording interface.
                self.checkPermissions(call)
            } else {
                call.reject("Record audio permission")
            }
        }

    }

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    @objc func initialize(_ call: CAPPluginCall) {
        hmsSDK = HMSSDK.build();
        call.resolve();
    }
    
    @objc func joinRoom(_ call: CAPPluginCall) {
        
        guard let id = call.options["userName"] as? String else {
           call.reject("Must provide a userName")
           return
         }
        
        guard let id = call.options["authToken"] as? String else {
           call.reject("Must provide an authToken")
           return
         }
        
        let userName = call.getString("userName") ?? "hello"
        let authToken = call.getString("authToken") ?? "token"
        let metaData = call.getString("metaData") ?? ""
        
        hmsConfig = HMSConfig(userName: userName, authToken: authToken, metadata: metaData)
        
        hmsSDK.join(config: hmsConfig, delegate: self)
        
        call.resolve();
    }
    
    @objc func muteAudio(_ call: CAPPluginCall) {
        hmsSDK.localPeer?.localAudioTrack()?.setMute(true);
        call.resolve();
    }
    
    @objc func unmuteAudio(_ call: CAPPluginCall) {
        hmsSDK.localPeer?.localAudioTrack()?.setMute(false);
        call.resolve();
    }
    
    @objc func muteVideo(_ call: CAPPluginCall) {
        hmsSDK.localPeer?.localVideoTrack()?.setMute(true);
        call.resolve();
    }
    
    @objc func unmuteVideo(_ call: CAPPluginCall) {
        hmsSDK.localPeer?.localVideoTrack()?.setMute(false);
        call.resolve();
    }
    
    @objc func enableSpeaker(_ call: CAPPluginCall) {
        try? hmsSDK.switchAudioOutput(to: .speaker);
        call.resolve();
    }
    
    @objc func disableSpeaker(_ call: CAPPluginCall) {
        try? hmsSDK.switchAudioOutput(to: .earpiece);
        call.resolve();
    }
    
    @objc func leaveRoom(_ call: CAPPluginCall) {
        hmsSDK.leave();
        call.resolve();
    }
    
}

extension HmsCapacitorPlugin: HMSUpdateListener {
    
    private func peerInterface(peer: HMSPeer) -> [String : Any] {
        let data = [
            "id": peer.peerID,
            "name": peer.name,
            "roleName": peer.role?.name ?? "",
            "isLocal": peer.isLocal,
            "isStarred": false,
            "auxiliaryTracks": peer.auxiliaryTracks ?? [],
            "customerUserId": peer.customerUserID ?? "",
            "metadata": peer.metadata ?? "",
            "joinedAt": peer.joinedAt
        ] as [String : Any]
        
        return data;
    }
    
    private func messageInterface(message: HMSMessage) -> [String : Any] {
        let data = [
            "id": message.time,
            "sender": message.sender?.peerID ?? "",
            "senderName": message.sender?.name ?? "",
            "senderUserId": message.sender?.customerUserID ?? "",
            "senderRole": message.sender?.role?.name ?? "",
            "recipientPeer": message.recipient.peerRecipient?.peerID ?? "",
            "time": message.time,
            "read": false,
            "type": message.type,
            "message": message.message,
            "ignored": false
        ] as [String : Any]
        
        return data;
    }

    
    public func on(join room: HMSRoom) {
        let roomData = [
            "id": room.roomID ?? "",
            "name": room.name ?? "",
            "isConnected": true,
            "peers": room.peers,
            "localPeer": room.peers[0],
            "roomState": room.rtmpStreamingState,
            "sessionId": room.sessionID ?? "",
            "startedAt": room.sessionStartedAt ?? "",
            "joinedAt": room.peers[0].joinedAt,
            "peerCount": room.peerCount ?? 0
        ] as [String : Any]
        self.notifyListeners("onJoin", data: [
            "room" : roomData
        ])
    }
    
    public func on(room: HMSRoom, update: HMSRoomUpdate) {
        //
    }
    
    public func on(peer: HMSPeer, update: HMSPeerUpdate) {
        self.notifyListeners("onPeerUpdate", data: [
            "hmsPeerUpdate": update,
            "hmsPeer": peerInterface(peer: peer)
        ])
    }
    
    public func on(track: HMSTrack, update: HMSTrackUpdate, for peer: HMSPeer) {
        self.notifyListeners("onTrackUpdate", data: [
            "hmsTrackUpdate": update,
            "hmsTrack": track,
            "hmsPeer": peerInterface(peer: peer)
        ])
    }
    
    public func on(error: Error) {
        self.notifyListeners("onError", data: [
            "error": error
        ])
    }
    
    public func on(message: HMSMessage) {
        self.notifyListeners("onMessageReceived", data: [
            "message": messageInterface(message: message)
        ])
    }
    
    public func on(updated speakers: [HMSSpeaker]) {
        self.notifyListeners("listenForDominantSpeaker", data: [
            "peer": peerInterface(peer: speakers[0].peer)
        ])
    }
    
    public func onReconnecting() {
        self.notifyListeners("onReconnecting", data: [
            "status": "reconnecting"
        ])
    }
    
    public func onReconnected() {
        self.notifyListeners("onReconnected", data: [
            "status": "reconnected"
        ])
    }
    
}

File: android/src/main/java/com/hms/capacitor/io/HmsCapacitorPlugin.java
package com.hms.capacitor.io;

import android.Manifest;
import android.app.Application;

import androidx.annotation.NonNull;

import com.getcapacitor.Bridge;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.google.gson.Gson;

import live.hms.video.audio.HMSAudioManager;
import live.hms.video.error.HMSException;
import live.hms.video.media.tracks.HMSLocalAudioTrack;
import live.hms.video.media.tracks.HMSLocalVideoTrack;
import live.hms.video.media.tracks.HMSTrack;
import live.hms.video.sdk.HMSActionResultListener;
import live.hms.video.sdk.HMSSDK;
import live.hms.video.sdk.HMSUpdateListener;
import live.hms.video.sdk.models.HMSConfig;
import live.hms.video.sdk.models.HMSLocalPeer;
import live.hms.video.sdk.models.HMSMessage;
import live.hms.video.sdk.models.HMSPeer;
import live.hms.video.sdk.models.HMSRemovedFromRoom;
import live.hms.video.sdk.models.HMSRoleChangeRequest;
import live.hms.video.sdk.models.HMSRoom;
import live.hms.video.sdk.models.enums.HMSPeerUpdate;
import live.hms.video.sdk.models.enums.HMSRoomUpdate;
import live.hms.video.sdk.models.enums.HMSTrackUpdate;
import live.hms.video.sdk.models.trackchangerequest.HMSChangeTrackStateRequest;

@CapacitorPlugin(name = "HmsCapacitor",  permissions = {
        @Permission(
                alias = "camera",
                strings = { Manifest.permission.CAMERA }
        ),
        @Permission(
                alias = "record_audio",
                strings = { Manifest.permission.RECORD_AUDIO }
        )
})
public class HmsCapacitorPlugin extends Plugin {

    private HmsCapacitor implementation = new HmsCapacitor();

    HMSSDK hmssdk;
    HMSConfig hmsConfig;

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }


    class MyHmsUpdateListener implements HMSUpdateListener {

        private JSObject RoomInterface (HMSRoom hmsRoom) {
            JSObject data = new JSObject();

            data.put("id", hmsRoom.getRoomId());
            data.put("name", hmsRoom.getName());
            data.put("isConnected", true);
            data.put("peers", hmsRoom.getPeerList().toArray());
            data.put("localPeer", hmsRoom.getLocalPeer().getPeerID());
            data.put("roomState", hmsRoom.getRtmpHMSRtmpStreamingState());
            data.put("sessionId", hmsRoom.getSessionId());
            data.put("startedAt", hmsRoom.getStartedAt());
            data.put("joinedAt", hmsRoom.getLocalPeer().getJoinedAt());
            data.put("peerCount", hmsRoom.getPeerCount());

            return data;
        }

        private JSObject MessageInterface (HMSMessage hmsMessage) {
            JSObject data = new JSObject();

            data.put("id", hmsMessage.getServerReceiveTime());
            data.put("sender", hmsMessage.getSender().getPeerID());
            data.put("senderName", hmsMessage.getSender().getName());
            data.put("senderUserId", hmsMessage.getSender().getCustomerUserID());
            data.put("senderRole", hmsMessage.getSender().getHmsRole().getName());
            data.put("recipientPeer", hmsMessage.getRecipient().getRecipientPeer().getPeerID());
            data.put("time", hmsMessage.getServerReceiveTime());
            data.put("read", false);
            data.put("type", hmsMessage.getType());
            data.put("message", hmsMessage.getMessage());
            data.put("ignored", false);

            return data;
        }

        private JSObject PeerInterface (HMSPeer hmsPeer) {
            JSObject data = new JSObject();

            data.put("id", hmsPeer.getPeerID());
            data.put("name", hmsPeer.getName());
            data.put("roleName", hmsPeer.getHmsRole().getName());
            data.put("isLocal", hmsPeer.isLocal());
            data.put("isStarred", false);
//            data.put("videoTrack", hmsPeer.getVideoTrack().getTrackId());
//            data.put("audioTrack", hmsPeer.getAudioTrack().getTrackId());
            data.put("auxiliaryTracks", hmsPeer.getAuxiliaryTracks());
            data.put("customerUserId", hmsPeer.getCustomerUserID());
            data.put("metadata", hmsPeer.getMetadata());
            data.put("joinedAt", hmsPeer.getJoinedAt());

            return data;
        }

        private JSObject ErrorInterface (HMSException e) {
            JSObject data = new JSObject();

            data.put("code", e.getCode());
            data.put("action", e.getAction());
            data.put("name", e.getName());
            data.put("message", e.getMessage());
            data.put("description", e.getDescription());
            data.put("isTerminal", e.isTerminal());

            return data;
        }


        @Override public void onJoin(@NonNull HMSRoom hmsRoom) {
            JSObject eventData = new JSObject();
            eventData.put("room", this.RoomInterface(hmsRoom));
            notifyListeners("onJoin", eventData);
        }
        @Override public void onMessageReceived(@NonNull HMSMessage hmsMessage) {
            JSObject eventData = new JSObject();
            eventData.put("message", this.MessageInterface(hmsMessage));
            notifyListeners("onMessageReceived", eventData);
        }
        @Override public void onPeerUpdate(@NonNull HMSPeerUpdate hmsPeerUpdate, @NonNull HMSPeer hmsPeer) {
            JSObject eventData = new JSObject();

            eventData.put("hmsPeerUpdate", hmsPeerUpdate);
            eventData.put("hmsPeer",  this.PeerInterface(hmsPeer));

            notifyListeners("onPeerUpdate", eventData);
        }
        @Override public void onReconnected() {
            JSObject eventData = new JSObject();
            eventData.put("status", "reconnected");
            notifyListeners("onReconnected", eventData);
        }
        @Override public void onReconnecting(@NonNull HMSException e) {
            JSObject eventData = new JSObject();
            eventData.put("status", "reconnecting");
            notifyListeners("onReconnecting", eventData);
        }
        @Override public void onRoleChangeRequest(@NonNull HMSRoleChangeRequest hmsRoleChangeRequest) {
            JSObject eventData = new JSObject();
            eventData.put("peer",  hmsRoleChangeRequest.getRequestedBy());
            notifyListeners("onRoleChange", eventData);
        }
        @Override public void onRoomUpdate(@NonNull HMSRoomUpdate hmsRoomUpdate, @NonNull HMSRoom hmsRoom) {}
        @Override public void onTrackUpdate(@NonNull HMSTrackUpdate hmsTrackUpdate, @NonNull HMSTrack hmsTrack, @NonNull HMSPeer hmsPeer) {
            JSObject eventData = new JSObject();

            if(hmsTrackUpdate == HMSTrackUpdate.TRACK_ADDED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_ADDED);
            } else if (hmsTrackUpdate == HMSTrackUpdate.TRACK_REMOVED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_REMOVED);
            } else if (hmsTrackUpdate == HMSTrackUpdate.TRACK_DEGRADED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_DEGRADED);
            } else if (hmsTrackUpdate == HMSTrackUpdate.TRACK_MUTED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_MUTED);
            } else if (hmsTrackUpdate == HMSTrackUpdate.TRACK_DESCRIPTION_CHANGED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_DESCRIPTION_CHANGED);
            } else if (hmsTrackUpdate == HMSTrackUpdate.TRACK_UNMUTED) {
                eventData.put("hmsTrackUpdate", HMSTrackUpdate.TRACK_UNMUTED);
            }

            eventData.put("hmsTrack",  hmsTrack);
            eventData.put("hmsPeer", this.PeerInterface(hmsPeer));

            notifyListeners("onTrackUpdate", eventData);
        }
        @Override public void onError(@NonNull HMSException e) {
            JSObject eventData = new JSObject();

            eventData.put("error",  this.ErrorInterface(e));

            notifyListeners("onError", eventData);
        }

        @Override
        public void onChangeTrackStateRequest(@NonNull HMSChangeTrackStateRequest hmsChangeTrackStateRequest) {

        }

        @Override
        public void onRemovedFromRoom(@NonNull HMSRemovedFromRoom hmsRemovedFromRoom) {

        }
    }


    @PluginMethod()
    public void initialize(PluginCall call) {
        this.hmssdk = new HMSSDK.Builder(this.getContext()).build();
        call.resolve();
    }

    @PluginMethod()
    public void joinRoom(PluginCall call) {
        String userName = call.getString("userName");
        String authToken = call.getString("authToken");
        String metaData = call.getString("metaData");

        if(!call.getData().has("userName")) {
            call.reject("UserName is required");
        }

        if(!call.getData().has("authToken")) {
            call.reject("authToken is required");
        }


        this.hmsConfig = new HMSConfig(userName, authToken, metaData);

        PermissionState audioPermissionState = getPermissionState("record_audio");

        if(audioPermissionState == PermissionState.GRANTED) {
            this.hmssdk.join(this.hmsConfig, new MyHmsUpdateListener());
        } else {
            requestPermissionForAlias("record_audio", call, "recordAudioPermsCallback");
        }

        call.resolve();
    }

    @PluginMethod()
    public void muteAudio(PluginCall call) {
        HMSLocalPeer myPeer = this.hmssdk.getLocalPeer();
        HMSLocalAudioTrack myAudioTrack = myPeer.getAudioTrack();
        if(myAudioTrack != null) {
            myAudioTrack.setMute(true);
        }
        call.resolve();
    }

    @PluginMethod()
    public void unmuteAudio(PluginCall call) {
        HMSLocalPeer myPeer = this.hmssdk.getLocalPeer();
        HMSLocalAudioTrack myAudioTrack = myPeer.getAudioTrack();
        if(myAudioTrack != null) {
            myAudioTrack.setMute(false);
        }
        call.resolve();
    }

    @PluginMethod()
    public void muteVideo(PluginCall call) {
        HMSLocalPeer myPeer = this.hmssdk.getLocalPeer();
        HMSLocalVideoTrack myVideoTrack = myPeer.getVideoTrack();
        if(myVideoTrack != null) {
            myVideoTrack.setMute(true);
        }
        call.resolve();
    }


    @PluginMethod()
    public void unmuteVideo(PluginCall call) {
        HMSLocalPeer myPeer = this.hmssdk.getLocalPeer();
        HMSLocalVideoTrack myVideoTrack = myPeer.getVideoTrack();
        if(myVideoTrack != null) {
            myVideoTrack.setMute(false);
        }
        call.resolve();
    }


    @PluginMethod()
    public void enableSpeaker(PluginCall call) {
        this.hmssdk.switchAudioOutput(HMSAudioManager.AudioDevice.SPEAKER_PHONE);
        call.resolve();
    }

    @PluginMethod()
    public void disableSpeaker(PluginCall call) {
        this.hmssdk.switchAudioOutput(HMSAudioManager.AudioDevice.AUTOMATIC);
        call.resolve();
    }


    class hmsActionResultListener implements HMSActionResultListener {

        @Override
        public void onSuccess() {

        }

        @Override
        public void onError(@NonNull HMSException e) {

        }
    }

    @PluginMethod()
    public void leaveRoom(PluginCall call) {
        this.hmssdk.leave(new hmsActionResultListener());
        call.resolve();
    }



    @PermissionCallback
    private void recordAudioPermsCallback(PluginCall call) {
        this.hmssdk.join(this.hmsConfig, new MyHmsUpdateListener());
    }
}
